### ISSUE TITLE
Use Pipfile instead of many requirements.txt files

### ISSUE THREAD

**Issue by ibnesayeed** [oduwsdl/ipwb#366]
Pipfile seems to be a more modern approach of maintaining dependencies https://github.com/pypa/pipfile

**Comment by machawk1** [comment]
Looks like base pip does not support the pipfile format. As far as I am aware, pipenv is not distributed with python like pip, so I think we ought to wait.

**Comment by ibnesayeed** [comment]
I don't think it is a problem, because it is a development environment setup related matter, it is not something that the user would be concerned.

**Comment by machawk1** [comment]
> it is a development environment setup related matter

Can we still recommend users install ipwb using `pip install ipwb` if we use a pipfile in lieu of various requirements files (as we currently do)?

**Comment by ibnesayeed** [comment]
`pip install ipwb` does not care about the development setup, it installs from the PyPi repo. Dependencies are described in the setup file for that. However, Pipfile allows having different set of dependencies in a single file instead of multiple requirements files. If I remember correctly, there are some commands in Pipfile to synchronize dependencies of setup file.

**Comment by machawk1** [comment]
@ibnesayeed So, a user will still be able to install using pip?

As I understand from the documentation, eventual support for using pipfiles from pip is planned.

**Comment by ibnesayeed** [comment]
> So, a user will still be able to install using pip?

Indeed!

**Comment by machawk1** [comment]
Ok, let's proceed on the efforts to use a pipfile then to resolve the variable, multiple requirements files.

**Comment by ibnesayeed** [comment]
I would suggest putting effort on this along with the migration to Py3 #51.

**Comment by ibnesayeed** [comment]
@anatoly-scherbakov we would like your thoughts on this thread and potentially moving to [poetry](https://python-poetry.org/) instead? We have not used poetry yet, but we certainly would like a better workflow of managing dependencies. Currently, we are making similar changes in two different files.

**Comment by machawk1** [comment]
As an update, I still could not find a reference stating that pip has native pipfile support yet.

As above, I believe it would be best to use a package manager included with Python rather than requiring the user to install a package manager (e.g., pipenv, poetry) prior to being able to install ipwb.

**Comment by ibnesayeed** [comment]
> As above, I believe it would be best to use a package manager included with Python rather than requiring the user to install a package manager (e.g., pipenv, poetry) prior to being able to install ipwb.

Users should be able to run `pip install ipwb` (or any other command they might be using to install Python packages) and get it running irrespective of what dependency management system we use, as long as we publish the package in PyPI. The changes will only affect those developers, not users, who want to tinker with the code. I think it is fair to enforce any development workflow that makes the life of maintainers easy and is not too peculiar for new contributors.

**Comment by machawk1** [comment]
> Users should be able to run pip install ipwb...irrespective of what dependency management system we use.

I agree. I am not seeing any indication that running `pip install ipwb` when using a pipfile to define the requirements is currently supported. poetry seems like a different dependency management system (I could be wrong). I do not see how moving to poetry would allow users to still type `pip install ipwb` and have ipwb installed on their system.

**Comment by ibnesayeed** [comment]
I think you are mixing a few things up. Running the command `pip install ipwb` has nothing to do with this repository and what code or files are present or absent in it here. As long as a well-bundled package is uploaded to PyPI with the name `ipwb` with its metadata and dependencies described in a format PyPI understands, users running `pip install ipwb` should have no problem installing it.

If you were to check the code out and from within the code directory run `pip install .`, then pip command will need the supporting metadata/dependency management files to function properly. However, this falls under the developer workflow, not the user workflow. Developers can be asked to use a different setup, if that makes life of maintainers easier.

Tools like "poetry" allow storing metadata and dependencies in a different file format, but when you package a bundle, they internally create necessary metadata files suitable for PyPI which is ready to upload. Better yet, we can use poetry itself to upload to PyPI. Currently, we are using a combination of setuptools, wheel, and twine to get all these things done.

**Comment by anatoly-scherbakov** [comment]
> @anatoly-scherbakov we would like your thoughts on this thread and potentially moving to [poetry](https://python-poetry.org/) instead? We have not used poetry yet, but we certainly would like a better workflow of managing dependencies. Currently, we are making similar changes in two different files.

I apologize for my taking so long to respond: there was a bit of a crunch time at work lately.

I agree with the comments above: indeed, it is not important what tool the package author used to create a pypi package; to do `pip install ipwb`, the user does not need to install Poetry.

I have not yet had a chance to publish a package to pypi with Poetry, but I believe this is rather straightforward - their main page https://python-poetry.org/ suggests it is as simple as typing `poetry publish`.

**Comment by machawk1** [comment]
Related aside, in [forking](https://github.com/machawk1/py-ipfs-http-client) and publishing py-ipfs-http-client to pypi, I used [flit](https://flit.readthedocs.io/) as preconfigured by the project and similar to how @anatoly-scherbakov mentioned with poetry, it was as simple as `flit publish` following the initial `flit init`, which read the parameters from the `pyproject.toml` in the root of the repo.
